PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc

```diff
@@ -0,0 +1,111 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+namespace pat {
+
+  class PackedCandidateTrackChi2Producer : public edm::stream::EDProducer<> {
+    typedef edm::ValueMap<float> FloatMap;
+
+  public:
+    explicit PackedCandidateTrackChi2Producer(const edm::ParameterSet& iConfig)
+        : candidateToken_(consumes<edm::View<pat::PackedCandidate>>(iConfig.getParameter<edm::InputTag>("candidates"))),
+          trackToken_(consumes<edm::View<reco::Track>>(iConfig.getParameter<edm::InputTag>("trackCollection"))),
+          doLostTracks_(iConfig.getParameter<bool>("doLostTracks")) {
+      if (doLostTracks_) {
+        track2LostTrackToken_ = consumes<edm::Association<pat::PackedCandidateCollection>>(
+            iConfig.getParameter<edm::InputTag>("candidates"));
+      } else {
+        candidate2PFToken_ =
+            consumes<edm::Association<reco::PFCandidateCollection>>(iConfig.getParameter<edm::InputTag>("candidates"));
+      }
+
+      produces<FloatMap>();
+    }
+    ~PackedCandidateTrackChi2Producer() override{};
+
+    void produce(edm::Event&, const edm::EventSetup&) override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    edm::EDGetTokenT<edm::View<pat::PackedCandidate>> candidateToken_;
+    edm::EDGetTokenT<edm::Association<reco::PFCandidateCollection>> candidate2PFToken_;
+    edm::EDGetTokenT<edm::Association<pat::PackedCandidateCollection>> track2LostTrackToken_;
+    edm::EDGetTokenT<edm::View<reco::Track>> trackToken_;
+    bool doLostTracks_;
```

Comment:
try to keep all/most parameters const where possible;
the current impl allows for candidateToken_, doLostTracks_, trackToken_.

Also, what is the reason to read edm::View?
There is some (small) overhead to construct it. In this case it looks like the most derived class objects are consumed anyways and a polymorphic access via a View seems not justified.